Msi 2365 get pickup locations endpoints refactoring

swnsma opened the pull request:
### Description (*)
Refactoring of Get Pickup Locations endpoint.
PR still in progress.
TODO: 
- ~~replace deprecated endpoints with single one~~
- ~~adjust tests~~
- ~~cover new endpoint with tests~~

Additional:
- ~~Rename `getSourceCode` to `getPickupLocationCode` for Pickup Location~~
- ~~Create single public constant for `pickup_location_code` and replace all private constants~~
- ~~Update API documentation~~

### Fixed Issues (if relevant)
1. https://github.com/magento/inventory/issues/2365: Get Pickup Locations endpoints refactoring

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Add draw api implementation.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Change Search Result to work with Search Request.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Make it works.
Adjust tests for distance-based search.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Remove old distance based search of Pickup Locations.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Improve structure sanity.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Improve check if filter exists for Address Filter resolver.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Remove deprecated API classes and replace it with usage of new service.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Replace Get Pickup Location API with SPI. Adjust tests. Fix typos.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Move Mapper from API to SPI as it does not provide any web interface.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Fix static tests and refactor the code.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Add integration tests coverage for Address Filter. Fix minor bugs.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Add integration general tests for Pickup Location filters, sort orders and paging

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Extract interface from Search Request Builder.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Add integration tests coverage for combined filters.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Fix static tests.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Add public constant for Pickup Location Code.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Rename getSourceCode to getPickupLocationCode.
Fix static tests.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Update comments for API classes.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Segregate Search Request Builder on separated filter builders.
Add integration tests coverage.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Revert support of Extension Attributes.
Need to be reverted as soon as https://github.com/magento/magento2/pull/24117 will be merged to mainline.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Fix static tests.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Update comments.

novikor commented:
![image](https://user-images.githubusercontent.com/26086561/63468415-80ee6880-c470-11e9-9afa-388183dd1997.png)
Please stop 🙅‍♂

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Remove rest usage of Search Request Extension.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Resolve issue with wrong total in search result.

novikor commented:
Is PR ready for review?

Commit:
Merge branch 'store-pickup' into MSI-2365-Get-Pickup-Locations-endpoints-refactoring

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Add translation for sort fields.

swnsma commented:
@novikor, yes, it is ready for review.

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Refactor extraction of distances and sort by distance for Pickup Locations.

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Add exception handling in case that entered address can not be recognized.

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Decrease number of API's.
Remove Distance, Address, Filter interfaces.

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Combine Address and General filters int Filter Set.
Remove dependency of Search Request on Filter.
Adjust tests.

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Fix static tests.

Commit:
Merge branch 'store-pickup' into MSI-2365-Get-Pickup-Locations-endpoints-refactoring

swnsma commented:
Adjusted according to the latest discussion:
- Address attributes, pickup location code and name have been merged to Filter Set sub-entity.

novikor commented on app/code/Magento/InventoryInStorePickup/Model/SearchResult/ExtractStrategy.php:
```diff
@@ -0,0 +1,52 @@
+     * @param DistanceBasedStrategy $distanceBasedStrategy
+     */
+    public function __construct(
+        DistanceBasedStrategy $distanceBasedStrategy
```
Is it just a decorator for `\Magento\InventoryInStorePickup\Model\SearchResult\DistanceBasedStrategy`? Why cannot we use it directly and remote this class?

swnsma replied:
I think something can be rethinked there.
For now we have: default extract and extract source, ordered by distance.
I think it is nice to have extension point there to have possibility to add new strategy for sources  extraction.

swnsma replied:
Discussed offline.
Refactored.
Please check.

novikor commented on app/code/Magento/InventoryInStorePickupApi/Api/Data/PickupLocationInterface.php:
```diff
@@ -18,26 +18,27 @@
 
     /**
-     * Get Pickup Location name
+     * Get Pickup Location Name.
```
Does it mean that we should use dot (`.`) for all the StorePickup PhpDocs now?

swnsma replied:
Yes.

novikor commented:
@magento run Integration Tests build

novikor commented:
@magento run Static tests build

novikor commented on app/code/Magento/InventoryInStorePickupApi/Model/GetPickupLocationInterface.php:
```diff
@@ -4,13 +4,13 @@
  */
 
-namespace Magento\InventoryInStorePickupApi\Api;
+namespace Magento\InventoryInStorePickupApi\Model;
```
Why?

swnsma replied:
This is nice sugar-service.
As API (inc webapi) we have `GetPickupLocations` service.

novikor commented on app/code/Magento/InventoryInStorePickupApi/Model/SearchRequestBuilderInterface.php:
```diff
@@ -0,0 +1,196 @@
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickupApi\Model;
```
I believe it should be moved to `Api`

swnsma replied:
I would like to have in `Api` folder only things related to webapi.
Moreover, in the `Framework` most of builders not even an `@api`.

novikor commented on app/code/Magento/InventoryInStorePickupApi/etc/webapi.xml:
```diff
@@ -7,20 +7,8 @@
-    </route>
-    <route url="/V1/inventory/in-store-pickup/pickup-location/:salesChannelType/:salesChannelCode/:pickupLocationCode" method="GET">
-        <service class="Magento\InventoryInStorePickupApi\Api\GetPickupLocationInterface" method="execute"/>
+    <route url="/V1/inventory/in-store-pickup/pickup-locations/" method="GET">
```
I believe there still is a probability of having a need to get the single pickup location by its code

swnsma replied:
You can get it with `GetPickupLocations` service by using filter by pickup location code.

novikor commented on app/code/Magento/InventoryInStorePickupShipping/Model/Carrier/Validation/SalesChannelValidator.php:
```diff
@@ -74,4 +79,23 @@ public function validate(RateRequest $rateRequest): ValidationResult
+     *
+     * @return bool
+     */
+    private function isAnyPickupLocationAvailable(string $websiteCode): bool
```
It would be cool to extract this method into a separate service. More likely it will not be the only place where we need this logic.

swnsma replied:
During refactoring I have found that this is the only place where this check is used.
I think we can add it in future if we need to.

novikor reviewed (changes requested):
Well, seems cool enough.
Please make sure all the required tests are green.

novikor commented:
![image](https://user-images.githubusercontent.com/26086561/64487865-a63cee00-d248-11e9-8e61-3fd4083fa6df.png)

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Edit phpdocs.

swnsma commented:
@novikor,
About Static Tests, I am not sure if something may be fixed there. 
I would like to have any suggestions.

About Integration tests, @ishakhsuvarov could you clarify if this is something related to PR or to the automation?

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Refactor Extractor Strategy.

swnsma commented:
![SearchRequest](https://user-images.githubusercontent.com/9420807/64522631-1b6cf980-d303-11e9-8314-7f1bf3fd6fcc.png)

novikor commented on app/code/Magento/InventoryInStorePickupApi/Model/SearchResult/Extractor.php:
```diff
@@ -0,0 +1,71 @@
+    {
+        foreach ($strategies as $strategy) {
+            if (!$strategy instanceof StrategyInterface) {
+                throw new \InvalidArgumentException(
```
Well, you could simplify `validateStrategies` and `getSources` just by returning `array_reduce` with type-hinted callback inside

novikor replied:
Okay 👌, I agree - it is cool approach for simplifying chains, but not the strategy.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Fix broken fieldset.

Commit:
Merge branch 'store-pickup' into MSI-2365-Get-Pickup-Locations-endpoints-refactoring

Commit:
Msi 2365 get pickup locations endpoints refactoring.
Refactor the code.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Fix tests after merge.
Intoruce Search Criteria Builder Decorator.

Commit:
MSI-2365-Get Pickup Locations endpoints refactoring.
Fix static tests.

swnsma commented:
I have introduced Decorator for Search Criteria Builder to isolate usage of not-public classes and satisfy static tests.

Commit:
Merge remote-tracking branch 'origin/store-pickup' into MSI-2365-Get-Pickup-Locations-endpoints-refactoring
